llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -634,11 +638,14 @@ AIEHazardRecognizer::getMemoryBanks(const MachineInstr *MI) const {
 
   const AIEBaseSubtarget &STI = AIEBaseSubtarget::get(*MI->getMF());
   const AIEBaseAddrSpaceInfo &ASI = STI.getAddrSpaceInfo();
-  MemoryBankBits MemoryBankUsed = ASI.getDefaultMemoryBank();
+  MemoryBankBits MemoryBankUsed = 0;
   for (auto &MMO : MI->memoperands()) {
+    unsigned AddrSpace = MMO->getAddrSpace();
     MemoryBankBits MemoryBank =
-        ASI.getMemoryBanksFromAddressSpace(MMO->getAddrSpace());
-    MemoryBankUsed &= MemoryBank;
+        AddrSpace                ? ASI.getMemoryBanksFromAddressSpace(AddrSpace)
+        : AddressSpaceNoneIsSafe ? 0
+                                 : ASI.getDefaultMemoryBank();
+    MemoryBankUsed |= MemoryBank;
```

Comment:
Note that this is more than an added option. We start at none, and collect the memory banks from all MMO's. We used to start at 'all banks' and intersect. The latter is incompatible with multi-memory ops like VLDB.4x